File: solutions/javascript/dfs/flattenNestedListIterator.js
class NestedIterator {
    /**
     * @constructor
     * @param {NestedInteger[]} nestedList
     */
    constructor(nestedList) {
        this.flattenedList = this.flattenList(nestedList);
        this.nextPosition = 0;
    }

    flattenList(nestedList) {
        const flattenedList = [];

        for (const nestedInteger of nestedList) {
            if (nestedInteger.isInteger()) {
                const integer = nestedInteger.getInteger();
                flattenedList.push(integer);
            } else {
                const nestedIntegerList = nestedInteger.getList();
                const flattenedNestedIntegerList =
                    this.flattenList(nestedIntegerList);

                for (const integer of flattenedNestedIntegerList) {
                    flattenedList.push(integer);
                }
            }
        }

        return flattenedList;
    }

    /**
     * @this NestedIterator
     * @returns {boolean}
     */
    hasNext() {
        return this.nextPosition < this.flattenedList.length;
    }

    /**
     * @this NestedIterator
     * @returns {integer}
     */
    next() {
        const integerToReturn = this.flattenedList[this.nextPosition];
        this.nextPosition++;

        return integerToReturn;
    }
}

File: solutions/python/dfs/flatten_nested_list_iterator.py
class NestedIterator:
    def __init__(self, nested_list: [NestedInteger]):
        self.flattened_list = self.flatten_list(nested_list)
        self.next_position = 0

    def flatten_list(self, nested_list):
        flattened_list = []

        for nested_integer in nested_list:
            if nested_integer.isInteger():
                integer = nested_integer.getInteger()
                flattened_list.append(integer)
            else:
                nested_integer_list = nested_integer.getList()
                flattened_nested_integer_list = self.flatten_list(nested_integer_list)

                for integer in flattened_nested_integer_list:
                    flattened_list.append(integer)

        return flattened_list

    def hasNext(self) -> bool:
        return self.next_position < len(self.flattened_list)
    
    def next(self) -> int:
        integer_to_return = self.flattened_list[self.next_position]
        self.next_position += 1
        
        return integer_to_return
